pkg/metrics/context_metrics.go

```diff
@@ -84,22 +84,73 @@ func (m ContextMetrics) Flush(timestamp float64) ([]*Serie, map[ckey.ContextKey]
 	errors := make(map[ckey.ContextKey]error)
 
 	for contextKey, metric := range m {
-		metricSeries, err := metric.flush(timestamp)
+		series = flushToSeries(
+			contextKey,
+			metric,
+			timestamp,
+			series,
+			errors)
+	}
 
-		if err == nil {
-			for _, serie := range metricSeries {
-				serie.ContextKey = contextKey
-				series = append(series, serie)
-			}
-		} else {
-			switch err.(type) {
-			case NoSerieError:
-				// this error happens in nominal conditions and shouldn't be returned
-			default:
-				errors[contextKey] = err
-			}
+	return series, errors
+}
+
+func flushToSeries(
+	contextKey ckey.ContextKey,
+	metric Metric,
+	bucketTimestamp float64,
+	series []*Serie,
+	errors map[ckey.ContextKey]error) []*Serie {
+	metricSeries, err := metric.flush(bucketTimestamp)
+
+	if err == nil {
+		for _, serie := range metricSeries {
+			serie.ContextKey = contextKey
+			series = append(series, serie)
+		}
+	} else {
+		switch err.(type) {
+		case NoSerieError:
+			// this error happens in nominal conditions and shouldn't be returned
+		default:
+			errors[contextKey] = err
 		}
 	}
+	return series
+}
 
-	return series, errors
+// aggregateContextMetricsByContextKey orders all Metric instances by context key,
+// representing the result as calls to the given callbacks.  The `callback` parameter
+// is called with each Metric in term, while `contextKeyChanged` is called after the
+// last Metric with each context key is processed. The last argument of the callback is the index
+// of the contextMetrics in contextMetricsCollection.
+//  For example:
+//     callback(key1, metric1, 0)
+//     callback(key1, metric2, 1)
+//     callback(key1, metric3, 2)
+//     contextKeyChanged()
+//     callback(key2, metric4, 0)
+//     contextKeyChanged()
+//     callback(key3, metric5, 0)
+//     callback(key3, metric6, 1)
+//     contextKeyChanged()
+func aggregateContextMetricsByContextKey(
```

Comment:
I think we should have a unit tests for this.

Reply:
It is already tested indirectly by
 https://github.com/DataDog/datadog-agent/blob/d3f531ae954a7435be0c5335e2fc76876b144102/pkg/metrics/context_metrics_flusher_test.go#L33

Let me know if you think it need a dedicated test.